File: samples/auction/auction.erl
%% auction.erl — Erlang reference implementation of the a4 auction demo.
%%
%% Idiomatic Erlang: gen_server per auction, record-based state,
%% pattern-matched handle_call/3. Compile and run with:
%%
%%   erlc auction.erl && erl -noshell -s auction run -s init stop

-module(auction).
-behaviour(gen_server).

-export([start_link/3, bid/3, close/1, leader/1, leading_bid/1, is_open/1,
         run/0]).
-export([init/1, handle_call/3, handle_cast/2]).

-record(state, {
    item           :: binary(),
    current_bid    :: integer(),
    highest_bidder :: binary(),
    status = open :: open | closed
}).

%%% API

start_link(Item, StartBid, StartBidder) ->
    gen_server:start_link(?MODULE, [Item, StartBid, StartBidder], []).

bid(Pid, Amount, Bidder) ->
    gen_server:call(Pid, {bid, Amount, Bidder}).

close(Pid) ->
    gen_server:call(Pid, close).

leader(Pid) ->
    gen_server:call(Pid, leader).

leading_bid(Pid) ->
    gen_server:call(Pid, leading_bid).

is_open(Pid) ->
    gen_server:call(Pid, is_open).

%%% gen_server callbacks

init([Item, StartBid, StartBidder]) ->
    {ok, #state{item = Item,
                current_bid = StartBid,
                highest_bidder = StartBidder}}.

handle_call({bid, _Amount, _Bidder}, _From, #state{status = closed} = S) ->
    {reply, false, S};
handle_call({bid, Amount, _Bidder}, _From,
            #state{current_bid = Current} = S) when Amount =< Current ->
    {reply, false, S};
handle_call({bid, Amount, Bidder}, _From, S) ->
    {reply, true, S#state{current_bid = Amount, highest_bidder = Bidder}};
handle_call(close, _From, S) ->
    {reply, ok, S#state{status = closed}};
handle_call(leader, _From, S) ->
    {reply, S#state.highest_bidder, S};
handle_call(leading_bid, _From, S) ->
    {reply, S#state.current_bid, S};
handle_call(is_open, _From, S) ->
    {reply, S#state.status =:= open, S}.

handle_cast(_Msg, S) ->
    {noreply, S}.

%%% Test driver

run() ->
    {ok, A} = start_link(<<"laptop">>, 100, <<>>),

    true  = bid(A, 150, <<"alice">>),
    <<"alice">> = leader(A),
    150   = leading_bid(A),

    false = bid(A, 120, <<"bob">>),      % too low
    <<"alice">> = leader(A),
    150   = leading_bid(A),

    true  = bid(A, 200, <<"bob">>),
    <<"bob">> = leader(A),
    200   = leading_bid(A),

    close(A),
    false = is_open(A),

    false = bid(A, 500, <<"carol">>),    % closed
    <<"bob">> = leader(A),
    200   = leading_bid(A),

    io:format("auction.erl: all assertions passed~n"),
    ok.

File: samples/auction/auction.exs
# auction.exs — Elixir reference implementation of the a4 auction demo.
#
# Idiomatic Elixir: GenServer per auction, struct-based state,
# pattern-matched handle_call clauses. Run with:
#
#   elixir auction.exs

defmodule Auction do
  use GenServer

  defstruct [:item, :current_bid, :highest_bidder, status: :open]

  # --- public API ---

  def start_link(item, start_bid, start_bidder) do
    GenServer.start_link(__MODULE__,
      %Auction{item: item, current_bid: start_bid, highest_bidder: start_bidder})
  end

  def bid(pid, amount, bidder), do: GenServer.call(pid, {:bid, amount, bidder})
  def close(pid),               do: GenServer.call(pid, :close)
  def leader(pid),              do: GenServer.call(pid, :leader)
  def leading_bid(pid),         do: GenServer.call(pid, :leading_bid)
  def open?(pid),               do: GenServer.call(pid, :is_open)

  # --- callbacks ---

  @impl true
  def init(state), do: {:ok, state}

  @impl true
  def handle_call({:bid, _amount, _bidder}, _from, %{status: :closed} = s),
    do: {:reply, false, s}

  def handle_call({:bid, amount, _bidder}, _from, %{current_bid: c} = s)
      when amount <= c,
    do: {:reply, false, s}

  def handle_call({:bid, amount, bidder}, _from, s),
    do: {:reply, true, %{s | current_bid: amount, highest_bidder: bidder}}

  def handle_call(:close, _from, s),
    do: {:reply, :ok, %{s | status: :closed}}

  def handle_call(:leader, _from, s),
    do: {:reply, s.highest_bidder, s}

  def handle_call(:leading_bid, _from, s),
    do: {:reply, s.current_bid, s}

  def handle_call(:is_open, _from, s),
    do: {:reply, s.status == :open, s}
end

# --- test driver ---

{:ok, a} = Auction.start_link("laptop", 100, "")

true  = Auction.bid(a, 150, "alice")
"alice" = Auction.leader(a)
150   = Auction.leading_bid(a)

false = Auction.bid(a, 120, "bob")      # too low
"alice" = Auction.leader(a)
150   = Auction.leading_bid(a)

true  = Auction.bid(a, 200, "bob")
"bob" = Auction.leader(a)
200   = Auction.leading_bid(a)

Auction.close(a)
false = Auction.open?(a)

false = Auction.bid(a, 500, "carol")    # closed
"bob" = Auction.leader(a)
200   = Auction.leading_bid(a)

IO.puts("auction.exs: all assertions passed")
